crates/sui-core/src/authority_active/cancellable_spawner.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::future::Future;
+use std::sync::Arc;
+use tokio::sync::{broadcast, OwnedRwLockWriteGuard, RwLock};
+use tokio::task::JoinHandle;
+use tracing::{debug, trace};
+
+use tap::TapFallible;
+
+use sui_types::error::{SuiError, SuiResult};
+
+/// CancellableSpawner allows spawned tasks to be cancelled easily without keeping track of every
+/// JoinHandle that gets created.
+pub struct CancellableSpawner {
+    cancel_tx: broadcast::Sender<()>,
+    spawn_lock: Arc<RwLock<()>>,
+    task_lock: Arc<RwLock<()>>,
```

Comment:
I do not like using Locks in this fashion, where they guard nothing. For the spawn lock i would probably create some inner struct that has the functionality to actually do the spawning so that the functionality is truly gated by the lock

Reply:
Refactored to move the task spawning inside the spawn lock. I find the new version of the code much less clear but it does at least make lock order mistakes impossible since the nesting enforces the lock order.

There's still nothing to put inside the innermost lock, but that's because its really just a more convenient alternative to an atomic task counter + a condvar.